src/replication/adapter/aliacr/adapter.go

```diff
@@ -157,42 +159,81 @@ func (a *adapter) FetchImages(filters []*model.Filter) (resources []*model.Resou
 	}
 	log.Debugf("FetchImages.repositories: %#v\n", repositories)
 
-	// list tags
-	for _, repository := range repositories {
-		var tags []string
-		tags, err = a.getTags(repository, client)
-		if err != nil {
-			return
-		}
+	rawResources := make([]*model.Resource, len(repositories))
+	var wg = new(sync.WaitGroup)
+	ctx, cancel := context.WithCancel(context.Background())
+	var passportsPool = utils.NewPassportsPool(adp.MaxConcurrency, ctx.Done())
 
-		var filterTags []string
-		if tagsPattern != "" {
-			for _, tag := range tags {
-				var ok bool
-				ok, err = util.Match(tagsPattern, tag)
-				if err != nil {
-					return
-				}
-				if ok {
-					filterTags = append(filterTags, tag)
+	for i, r := range repositories {
+		wg.Add(1)
+		go func(index int, repo aliRepo) {
+			defer func() {
+				wg.Done()
+			}()
+
+			// Return false means no passport acquired, and no valid passport will be dispatched any more.
+			// For example, some crucial errors happened and all tasks should be cancelled.
+			if ok := passportsPool.Apply(); !ok {
+				return
+			}
+			defer func() {
+				passportsPool.Revoke()
+			}()
+
+			var tags []string
+			tags, err = a.getTags(repo, client)
+			if err != nil {
+				log.Errorf("List tags for repo '%s' error: %v", repo.RepoName, err)
+				cancel()
+				return
+			}
+
+			var filterTags []string
+			if tagsPattern != "" {
+				for _, tag := range tags {
+					var ok bool
+					ok, err = util.Match(tagsPattern, tag)
+					if err != nil {
+						log.Errorf("Match tag '%s' error: %v", tag, err)
+						cancel()
+						return
+					}
+					if ok {
+						filterTags = append(filterTags, tag)
+					}
 				}
+			} else {
+				filterTags = tags
 			}
-		} else {
-			filterTags = tags
-		}
 
-		if len(filterTags) > 0 {
-			resources = append(resources, &model.Resource{
-				Type:     model.ResourceTypeImage,
-				Registry: a.registry,
-				Metadata: &model.ResourceMetadata{
-					Repository: &model.Repository{
-						Name: filepath.Join(repository.RepoNamespace, repository.RepoName),
+			if len(filterTags) > 0 {
+				rawResources[index] = &model.Resource{
+					Type:     model.ResourceTypeImage,
+					Registry: a.registry,
+					Metadata: &model.ResourceMetadata{
+						Repository: &model.Repository{
+							Name: filepath.Join(repo.RepoNamespace, repo.RepoName),
+						},
+						Vtags:  filterTags,
+						Labels: []string{},
 					},
-					Vtags:  filterTags,
-					Labels: []string{},
-				},
-			})
+				}
+			} else {
+				rawResources[index] = nil
+			}
+		}(i, r)
+	}
+	wg.Wait()
+
+	err = ctx.Err()
+	cancel()
```

Comment:
seems this `cancel` can be deferred too?

Reply:
Yes